File: lib/models/filter_rules.dart
class FilterRules {
  bool autoAcceptEnabled;
  double minFare;
  double maxPickupKm;
  double minTripKm;
  double minFarePerKm;
  String blacklistKeywords; // comma separated
  String whitelistKeywords; // comma separated
  int acceptDelayMs;

  FilterRules({
    this.autoAcceptEnabled = false,
    this.minFare = 0,
    this.maxPickupKm = 5,
    this.minTripKm = 0,
    this.minFarePerKm = 0,
    this.blacklistKeywords = '',
    this.whitelistKeywords = '',
    this.acceptDelayMs = 800,
  });

  factory FilterRules.fromMap(Map<dynamic, dynamic> map) {
    return FilterRules(
      autoAcceptEnabled: map['autoAcceptEnabled'] as bool? ?? false,
      minFare: (map['minFare'] as num?)?.toDouble() ?? 0,
      maxPickupKm: (map['maxPickupKm'] as num?)?.toDouble() ?? 5,
      minTripKm: (map['minTripKm'] as num?)?.toDouble() ?? 0,
      minFarePerKm: (map['minFarePerKm'] as num?)?.toDouble() ?? 0,
      blacklistKeywords: map['blacklistKeywords'] as String? ?? '',
      whitelistKeywords: map['whitelistKeywords'] as String? ?? '',
      acceptDelayMs: (map['acceptDelayMs'] as int?) ?? 800,
    );
  }

  Map<String, dynamic> toMap() => {
        'autoAcceptEnabled': autoAcceptEnabled,
        'minFare': minFare,
        'maxPickupKm': maxPickupKm,
        'minTripKm': minTripKm,
        'minFarePerKm': minFarePerKm,
        'blacklistKeywords': blacklistKeywords,
        'whitelistKeywords': whitelistKeywords,
        'acceptDelayMs': acceptDelayMs,
      };

  List<String> get blacklistList =>
      blacklistKeywords.split(',').map((e) => e.trim()).where((e) => e.isNotEmpty).toList();

  List<String> get whitelistList =>
      whitelistKeywords.split(',').map((e) => e.trim()).where((e) => e.isNotEmpty).toList();
}

File: android/app/src/main/kotlin/com/example/uber_surge_map_hk/model/FilterRules.kt
package com.example.uber_surge_map_hk.model

data class FilterRules(
    var autoAcceptEnabled: Boolean = false,
    var minFare: Double = 0.0,
    var maxPickupKm: Double = 5.0,
    var minTripKm: Double = 0.0,
    var minFarePerKm: Double = 0.0,
    var blacklistKeywords: String = "",
    var whitelistKeywords: String = "",
    var acceptDelayMs: Int = 800
) {
    val blacklistList: List<String>
        get() = blacklistKeywords.split(",").map { it.trim() }.filter { it.isNotEmpty() }

    val whitelistList: List<String>
        get() = whitelistKeywords.split(",").map { it.trim() }.filter { it.isNotEmpty() }

    fun passes(order: OrderModel): Boolean {
        if (minFare > 0 && order.fare > 0 && order.fare < minFare) return false
        if (maxPickupKm > 0 && order.pickupDistanceKm > 0 && order.pickupDistanceKm > maxPickupKm) return false
        if (minTripKm > 0 && order.tripDistanceKm > 0 && order.tripDistanceKm < minTripKm) return false
        if (minFarePerKm > 0 && order.farePerKm > 0 && order.farePerKm < minFarePerKm) return false

        val fullAddr = "${order.pickupAddress} ${order.destinationAddress}"
        if (blacklistList.any { fullAddr.contains(it) }) return false
        if (whitelistList.isNotEmpty() && whitelistList.none { fullAddr.contains(it) }) return false

        return true
    }
}
